internal/vault/configmap.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/hashicorp/vault-secrets-operator/internal/common"
+)
+
+type ShutDownMode int
+
+type ShutDownStatus int
+
+const (
+	ShutDownModeUnknown ShutDownMode = iota
+	ShutDownModeRevoke
+	ShutDownModeNoRevoke
+
+	ConfigMapKeyShutDownMode   = "shutDownMode"
+	ConfigMapKeyShutDownStatus = "shutDownStatus"
+
+	ShutDownStatusDone ShutDownStatus = iota
+	ShutDownStatusFailed
+	ShutDownStatusPending
+	ShutDownStatusUnknown
+)
+
+func (m ShutDownMode) String() string {
+	switch m {
+	case ShutDownModeRevoke:
+		return "revoke"
+	case ShutDownModeNoRevoke:
+		return "no-revoke"
+	default:
+		return "unknown"
+	}
+}
+
+func (s ShutDownStatus) String() string {
+	switch s {
+	case ShutDownStatusPending:
+		return "pending"
+	case ShutDownStatusFailed:
+		return "failed"
+	case ShutDownStatusDone:
+		return "done"
+	default:
+		return "unknown"
+	}
+}
+
+type OnConfigMapChange func(context.Context, client.Client, *corev1.ConfigMap)
+
+func WaitForManagerConfigMapModified(ctx context.Context, watcher watch.Interface, c client.Client, onChanges ...OnConfigMapChange) {
+	defer watcher.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event := <-watcher.ResultChan():
+			if event.Type == watch.Modified {
+				if cm, ok := event.Object.(*corev1.ConfigMap); ok {
+					for _, onChange := range onChanges {
+						onChange(ctx, c, cm)
+					}
+				}
+			}
+		}
+	}
+}
+
+// getManagerConfigMapList returns the manager configmap list that has configmaps with labels matching
+// "app.kubernetes.io/component": "controller-manager". For simplicity, we assume there should be one manager configmap
+func getManagerConfigMapList(ctx context.Context, c client.Client) (*corev1.ConfigMapList, error) {
+	labels := client.MatchingLabels{"app.kubernetes.io/component": "controller-manager"}
+	opts := []client.ListOption{
+		client.InNamespace(common.OperatorNamespace),
+		labels,
+	}
+
+	var list corev1.ConfigMapList
+	if err := c.List(ctx, &list, opts...); err != nil {
+		return nil, err
+	}
+
+	if len(list.Items) == 0 {
+		return nil, fmt.Errorf("no configmaps matching labels=%v found in the operator namespace", labels)
+	}
+
+	if len(list.Items) > 1 {
```

Comment:
I'm wondering if it make sense to not return error in this case, and just log that we found more than one and return the first element so that we can proceed still? It looks like we're only processing element `0` anyways in the calling function. Thoughts?

Reply:
I removed it the len > 1 error. Not sure if the case there are more than 1 manager ConfigMap existing in the operator. I can put a note/TODO for now. We can reiterate in the next PR?